test/unit/puppetlabs/services/jruby/jruby_pool_test.clj

```diff
@@ -99,6 +99,62 @@
               (doseq [k (keys counts)]
                 (is (= (inc (counts k)) (new-counts k)))))))))))
 
+(deftest borrow-while-pool-is-being-initialized-test
+  (testing "borrow will block until an instance is available while the pool is coming online"
+    (let [pool-initialized? (promise)
+          init-fn (fn [instance] @pool-initialized? instance)
+          pool-size 1
+          config (jruby-testutils/jruby-config
+                  {:max-active-instances pool-size
+                   :lifecycle {:initialize-pool-instance init-fn}})
+          pool-context (jruby-core/create-pool-context config)]
+
+      ;; start a pool initialization, which will block on the `initialize-pool-instance`
+      ;; function's deref of the promise
+      (jruby-agents/send-prime-pool! pool-context)
+
+      ;; start a borrow, which should block until an instance becomes available
+      (let [borrow-instance (future (jruby-core/borrow-from-pool-with-timeout
+                                     pool-context
+                                     :borrow-during-pool-init-test
+                                     []))]
+
+        (is (not (realized? borrow-instance)))
+
+        ;; deliver the promise, allowing the pool initialization to complete
+        (deliver pool-initialized? true)
+
+        ;; now the borrow can complete
+        (is (some? @borrow-instance))))))
```

Comment:
I feel like at one point (maybe on another team or something) I was told or heard the recommendation to use `some?` instead of `not nil?`. But agreed that asserting that it's actually an instance is a better test.